Fix/default ref frame

martiniil opened the pull request:
Targets #102
- [x] update changelog
- [x] increase API version

martiniil committed:
Increase test coverage by moving prbt w.r.t. world

* use prbt_base as default reference frame in test
* expect that the requests contain poses in the root frame (world)

martiniil committed:
Fix transformation of poses in robot api. Fixes #102

gavanderhoorn commented on pilz_robot_programming/src/pilz_robot_programming/commands.py:
```diff
@@ -273,17 +273,18 @@ def _get_goal_pose(self, robot):
-            return self._goal
-        else:
-            return Pose(position=self._goal.position, orientation=current_pose.orientation)
+        identity_quat = Quaternion()
```
```suggestion
        identity_quat = Quaternion(w=1.0)
```

martiniil replied:
Thanks! @gavanderhoorn

jschleicher reviewed (commented):
Where is the actual change? I cannot see from the diff, where this fixes #102. Could you explain, please?

What about documentation and changelog? Does this change the behaviour and make a `ROBOT_API ` change necessary?

martiniil commented:
These lines introduced the bug:

        if not self._reference_frame == _DEFAULT_BASE_LINK:
            return _to_robot_reference(robot, self._reference_frame, self._goal)

The actual change is that now the goal is always transformed to the root frame of the robot.

martiniil commented:
It does change the behaviour of the API. Now it behaves as promised by the API docu.

jschleicher commented on pilz_robot_programming/src/pilz_robot_programming/commands.py:
```diff
@@ -640,9 +640,6 @@ def _to_robot_reference(robot, pose_frame, goal_pose_custom_ref):
 
-    if not _is_quaternion_initialized(goal_pose_custom_ref.orientation):
-        goal_pose_custom_ref.orientation.w = 1
-
```
Would it be feasible to delegate this task to moveit? Simply set the `header.frame_id` in your request and leave the transformation to the motion planner. As far as I understand, it should be implemented there (as well) and thus is some code duplication...

martiniil replied:
Sounds good. I will give it a try.

gavanderhoorn replied:
> Would it be feasible to delegate this task to moveit?

While in general I would agree with reusing as much as possible, in this case it might not be such a bad thing to try and make sure that things under control of the planner / robot programming framework are consistent.

It's a small thing to make sure that all outgoing planning requests / trajectories / poses are in robot base frame. It also makes the software in this repository more robust against changes in MoveIt (it might stop doing what @jschleicher writes in his comment) and if there is a planning framework that is not MoveIt, but provides the same/a similar interface, but does not transform poses into robot base frame, bad things can happen.

I'm a rather defensive programmer, so this may not be compatible with the style of the rest of `pilz_robot_programming`.

martiniil replied:
@jschleicher @gavanderhoorn Thanks for your inputs. I don't have a strong opinion about it. I would keep it as it is, unless more reviewers request a change.

agutenkunst replied:
@gavanderhoorn thank you for your input!
I also agree with @jschleicher that the motion planner should be able to handle other frame_ids but maybe we should implement this independent from the need of the underlying PythonAPI.

jschleicher replied:
Ok, fine. Keep this for now.

jschleicher commented:
Okay, citing @agutenkunst (see [#102-issuecomment-467816895](https://github.com/PilzDE/pilz_industrial_motion/issues/102#issuecomment-467816895)) this requires changing the `ROBOT_API`, since the custom python code could move the robot differently after this change.

martiniil committed:
Increase robot API version

The changes introduced in commit d50bf76 could make the robot move differently.

agutenkunst commented on pilz_robot_programming/test/integrationtests/tst_api_cmd_conversion.py:
```diff
@@ -29,7 +29,8 @@
 TARGET_LINK_NAME = "prbt_tcp"
-API_VERSION = "1"
+BASE_LINK_NAME = "prbt_base"
+API_VERSION = "2"
```
I would like to open the discussion if we support backwards compatibility to some degree...
Any opinion? To much effort?

jschleicher replied:
+1 for allowing execution of API=1 programs (with former behaviour) and a deprecation warning

jschleicher commented:
@agutenkunst  @martiniil any progress?

martiniil commented:
@jschleicher We discussed several concepts for a backwards compability and scheduled its implementation. Until then, merging this PR is postponed.
